### PR TITLE
GC+ delete expired tokens

### LINKED ISSUE #4940: GC+ delete expired address tokens
part of #4476

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #4940

### CODE REVIEW

File: cmd/lakefs/cmd/run.go [hunk 1]
```diff
@@ -398,6 +401,55 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+	s.StartBlocking()
+}
+
+func deleteExpiredAddressTokens(ctx context.Context, c *catalog.Catalog) {
+	const (
+		jobWorkers = 3
+		size       = 10
+	)
+	rChan := make(chan *catalog.Repository, size)
+	for i := 0; i < jobWorkers; i++ {
```
Comment (N-o-Z) [inline cmd/lakefs/cmd/run.go:422]: Is there a specific reason why we want to do it concurrently?
As it seems, there's no real significance for the time it takes for this job to run

File: cmd/lakefs/cmd/run.go [hunk 2]
```diff
@@ -398,6 +401,55 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
 	}
 }
 
+func deleteExpiredAddressesJob(ctx context.Context, c *catalog.Catalog) {
+	s := gocron.NewScheduler(time.UTC)
+	_, err := s.Every(1).Hour().Do(func() {
```
Comment (N-o-Z) [inline cmd/lakefs/cmd/run.go:406]: I think the duration should be tightly coupled with the token expiration time parameter (for example: 2 times the expiration time)

File: cmd/lakefs/cmd/run.go [hunk 3]
```diff
@@ -398,6 +401,55 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+		for _, repo := range repos {
+			rChan <- repo
+			next = repo.Name
+		}
+	}
+}
+
+func rWorker(ctx context.Context, c *catalog.Catalog, rChan <-chan *catalog.Repository) {
+	for r := range rChan {
+		err := c.DeleteExpiredAddressTokens(ctx, r.Name)
```
Comment (N-o-Z) [inline cmd/lakefs/cmd/run.go:446]: I think this function needs to be as part of this PR, as far as I understand only this flow should invoke DeleteExpiredAddressTokens

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks for the changes.
Missing unit testing on this code, mainly:
1. Sanity
2. Test with no repositories
3. Test with repository with no tokens
4. Test repository deleted before calling delete

**Review by itaiad200 — CHANGES_REQUESTED** [review]

File: cmd/lakefs/cmd/run.go (hunk 2 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go]: done

File: cmd/lakefs/cmd/run.go (hunk 3 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go]: done

File: cmd/lakefs/cmd/run.go [hunk 4]
```diff
@@ -398,6 +401,55 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
 	}
 }
 
+func deleteExpiredAddressesJob(ctx context.Context, c *catalog.Catalog) {
+	s := gocron.NewScheduler(time.UTC)
+	_, err := s.Every(1).Hour().Do(func() {
+		deleteExpiredAddressTokens(ctx, c)
+	})
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:409]: done

**Comment by eden-ohana** [comment]
> Thanks for the changes. Missing unit testing on this code, mainly:
> 
> 1. Sanity
> 2. Test with no repositories
> 3. Test with repository with no tokens
> 4. Test repository deleted before calling delete

#4954

File: cmd/lakefs/cmd/run.go [hunk 5]
```diff
@@ -14,8 +14,11 @@ import (
 	"syscall"
 	"time"
 
+	"github.com/treeverse/lakefs/pkg/graveler/ref"
+
```
Comment (N-o-Z) [inline cmd/lakefs/cmd/run.go:18]: Redundant newline

File: pkg/graveler/ref/manager.go [hunk 6]
```diff
@@ -606,3 +606,24 @@ func (m *Manager) resolveAddressTokenTime(address string) (time.Time, error) {
...
+func (m *Manager) DeleteExpiredAddressTokens(ctx context.Context, repository *graveler.RepositoryRecord) error {
+	itr, err := m.ListAddressTokens(ctx, repository)
+	if err != nil {
+		return nil
+	}
+	defer itr.Close()
+	for itr.Next() {
+		token := itr.Value()
+		if m.IsTokenExpired(token) != nil {
+			err := m.kvStore.DeleteMsg(ctx, graveler.RepoPartition(repository), []byte(graveler.LinkedAddressPath(token.Address)))
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager.go:619]: Use m.deleteAddressToken

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Sorry forgot to request change.
Some minor comments

File: pkg/graveler/ref/manager.go [hunk 7]
```diff
@@ -606,3 +606,24 @@ func (m *Manager) resolveAddressTokenTime(address string) (time.Time, error) {
...
+
+func (m *Manager) DeleteExpiredAddressTokens(ctx context.Context, repository *graveler.RepositoryRecord) error {
+	itr, err := m.ListAddressTokens(ctx, repository)
+	if err != nil {
+		return nil
+	}
+	defer itr.Close()
+	for itr.Next() {
+		token := itr.Value()
+		if m.IsTokenExpired(token) != nil {
```
Comment (nopcoder) [inline pkg/graveler/ref/manager.go:618]: Why it doesn't return bool? if error can be more than just expired - it means we can't compare with != nil here.

File: pkg/graveler/ref/manager.go [hunk 8]
```diff
@@ -606,3 +606,24 @@ func (m *Manager) resolveAddressTokenTime(address string) (time.Time, error) {
...
+			err := m.kvStore.DeleteMsg(ctx, graveler.RepoPartition(repository), []byte(graveler.LinkedAddressPath(token.Address)))
+			if err != nil {
+				return nil
+			}
```
Comment (nopcoder) [inline pkg/graveler/ref/manager.go:622]: We don't return the error and don't log it.

**Review by nopcoder — CHANGES_REQUESTED** [review]
Added my comments.
Also - the test should render something that will look like the current generated path - it looks like we return and set xid. I may be behind in one PR to understand why.
Suggest (maybe in a different PR) to address the naming - we call it 'address tokens', but we actually mean different type of address we generated. Very confusing to manage tokens when we reference physical paths.

File: cmd/lakefs/cmd/run.go (hunk 1 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go]: https://github.com/treeverse/lakeFS/issues/4959

File: cmd/lakefs/cmd/run.go (hunk 5 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:18]: done

File: pkg/graveler/ref/manager.go (hunk 6 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:619]: done

**Review by N-o-Z — APPROVED** [review]
LGTM! Approved with one comment

File: cmd/lakefs/cmd/run.go [hunk 9]
```diff
@@ -395,6 +400,40 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+		var repos []*catalog.Repository
+		repos, hasMore, err = c.ListRepositories(ctx, -1, "", next)
+		if err != nil {
+			logging.Default().WithError(err).Debug("failed to list repositories while deleting expired tokens")
+		}
+
+		for _, repo := range repos {
+			err := c.DeleteExpiredAddressTokens(ctx, repo.Name)
+			if err != nil {
+				logging.Default().WithError(err).WithField("repository", repo.Name).Debug("failed to delete expired address tokens")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:430]: done

File: pkg/graveler/ref/manager.go (hunk 8 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:627]: done

File: cmd/lakefs/cmd/run.go [hunk 10]
```diff
@@ -395,6 +400,40 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+func deleteExpiredAddressTokens(ctx context.Context, c *catalog.Catalog) {
+	hasMore := true
+	next := ""
+
+	for hasMore {
+		var err error
+		var repos []*catalog.Repository
+		repos, hasMore, err = c.ListRepositories(ctx, -1, "", next)
+		if err != nil {
+			logging.Default().WithError(err).Debug("failed to list repositories while deleting expired tokens")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:424]: done

File: cmd/lakefs/cmd/run.go [hunk 11]
```diff
@@ -395,6 +400,40 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+}
+
+func deleteExpiredAddressTokens(ctx context.Context, c *catalog.Catalog) {
+	hasMore := true
+	next := ""
+
+	for hasMore {
+		var err error
+		var repos []*catalog.Repository
+		repos, hasMore, err = c.ListRepositories(ctx, -1, "", next)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:422]: done

File: cmd/lakefs/cmd/run.go [hunk 12]
```diff
@@ -395,6 +400,40 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
...
+		deleteExpiredAddressTokens(ctx, c)
+	})
+	if err != nil {
+		logging.Default().WithError(err).Debug("failed execute delete expired addresses job")
+	} else {
+		s.StartBlocking()
+	}
+}
+
+func deleteExpiredAddressTokens(ctx context.Context, c *catalog.Catalog) {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:415]: done

File: cmd/lakefs/cmd/run.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:422]: done

File: cmd/lakefs/cmd/run.go [hunk 13]
```diff
@@ -191,6 +192,8 @@ var runCmd = &cobra.Command{
 		}
 		defer func() { _ = c.Close() }()
 
+		go deleteExpiredAddressesJob(ctx, c)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:195]: done

File: cmd/lakefs/cmd/run.go [hunk 14]
```diff
@@ -395,6 +400,40 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
 	}
 }
 
+func deleteExpiredAddressesJob(ctx context.Context, c *catalog.Catalog) {
+	s := gocron.NewScheduler(time.UTC)
+
+	_, err := s.Every(ref.AddressTokenTime).Do(func() {
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:405]: done

File: cmd/lakefs/cmd/run.go [hunk 15]
```diff
@@ -395,6 +404,25 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
 	}
 }
 
+func deleteExpiredAddressesJob(ctx context.Context, s *gocron.Scheduler, c *catalog.Catalog) error {
+	const deleteExpiredAddressTokensPeriod = 3
+	job, err := s.Every(deleteExpiredAddressTokensPeriod * ref.AddressTokenTime).Do(func() {
+		err := c.DeleteExpiredAddressTokens(ctx)
```
Comment (nopcoder) [inline cmd/lakefs/cmd/run.go:410]: Not sure this method should return an error if it returns an error just when it fails to list repositories.
- Change to no value in return and log the error inside the call - like other calls done inside
- Change the function implementation and return multi error to list all the repositories we failed to expire.
I think option 1 is better for this case.

File: cmd/lakefs/cmd/run.go [hunk 16]
```diff
@@ -395,6 +404,25 @@ func checkRepos(ctx context.Context, logger logging.Logger, authMetadataManager
 	}
 }
 
+func deleteExpiredAddressesJob(ctx context.Context, s *gocron.Scheduler, c *catalog.Catalog) error {
```
Comment (nopcoder) [inline cmd/lakefs/cmd/run.go:407]:
```suggestion
func scheduleExpiredAddressesJob(ctx context.Context, s *gocron.Scheduler, c *catalog.Catalog) error {
```

File: pkg/graveler/ref/manager.go [hunk 17]
```diff
@@ -606,3 +607,28 @@ func (m *Manager) resolveAddressTokenTime(address string) (time.Time, error) {
 	}
 	return id.Time(), nil
 }
+
+func (m *Manager) DeleteExpiredAddressTokens(ctx context.Context, repository *graveler.RepositoryRecord) error {
+	itr, err := m.ListAddressTokens(ctx, repository)
+	if err != nil {
+		return nil
+	}
```
Comment (nopcoder) [inline pkg/graveler/ref/manager.go:615]: return the error

File: pkg/graveler/ref/manager.go [hunk 18]
```diff
@@ -606,3 +607,28 @@ func (m *Manager) resolveAddressTokenTime(address string) (time.Time, error) {
...
+			err := m.deleteAddressToken(ctx, repository, token.Address)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	if err = itr.Err(); err != nil {
+		return err
+	}
+	return nil
```
Comment (nopcoder) [inline pkg/graveler/ref/manager.go:633]:
```suggestion
	return itr.Err()
```

File: pkg/graveler/ref/manager_test.go [hunk 19]
```diff
@@ -1111,3 +1111,49 @@ func TestManager_IsTokenExpired(t *testing.T) {
 		t.Fatalf("expected token expired")
 	}
 }
+
+func TestManager_DeleteExpiredAddressTokens(t *testing.T) {
+	r, _ := testRefManager(t)
+	ctx := context.Background()
+	repository, err := r.CreateRepository(ctx, "repo1", graveler.Repository{
```
Comment (nopcoder) [inline pkg/graveler/ref/manager_test.go:1118]: Question - we create a real repository, but use fake ref manager? if the answer is yes - we should just initialize repository record filled with information and not call the real API to create a repo.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/graveler/ref/manager_test.go [hunk 20]
```diff
@@ -1111,3 +1111,49 @@ func TestManager_IsTokenExpired(t *testing.T) {
...
+		DefaultBranchID:  "main",
+	})
+	testutil.Must(t, err)
+
+	a := xid.NewWithTime(time.Now()).String()
+	b := xid.NewWithTime(time.Now().Add(-10 * time.Hour)).String() // expired
+	c := xid.NewWithTime(time.Now().Add(-7 * time.Hour)).String()  // expired
+	prefix := "data/skdgjslk/"
+
+	tokens := []string{prefix + a, prefix + b, prefix + c}
```
Comment (nopcoder) [inline pkg/graveler/ref/manager_test.go:1130]: Can we have different prefixes - similar to what we have in real time (reverse order)

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: cmd/lakefs/cmd/run.go (hunk 16 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:407]: done

File: pkg/graveler/ref/manager.go (hunk 17 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:615]: done

File: pkg/graveler/ref/manager.go (hunk 18 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:633]: done

File: pkg/graveler/ref/manager_test.go (hunk 20 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager_test.go:1130]: done

File: pkg/graveler/ref/manager_test.go (hunk 19 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager_test.go:1118]: we use real ref manager

File: cmd/lakefs/cmd/run.go (hunk 15 above)
Comment (eden-ohana) [reply cmd/lakefs/cmd/run.go:410]: done

File: pkg/graveler/ref/manager.go [hunk 21]
```diff
@@ -593,7 +594,7 @@ func (m *Manager) IsTokenExpired(token *graveler.LinkAddressData) (bool, error)
 	}
 	expiry := creationTime.Add(AddressTokenTime)
 	if expiry.Before(time.Now()) {
```
Comment (nopcoder) [inline pkg/graveler/ref/manager.go:596]: Can address in a different PR - easy to read:

```go
if time.Since(creationTime) > AddressTokenTime {
```

**Review by nopcoder — APPROVED** [review]
Waiting for the other PR that will cleanup all the 'Token' from the code.
Left couple of optional comments.

File: pkg/catalog/catalog.go [hunk 22]
```diff
@@ -1949,6 +1949,33 @@ func (c *Catalog) VerifyAddressToken(ctx context.Context, repository, token stri
 	return c.Store.VerifyAddressToken(ctx, repo, token)
 }
 
+func (c *Catalog) DeleteExpiredAddressTokens(ctx context.Context) {
+	it, err := c.Store.ListRepositories(ctx)
+	if err != nil {
+		c.log.WithError(err).Warn("failed to list repositories")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1955]: done

File: pkg/catalog/catalog.go [hunk 23]
```diff
@@ -1949,6 +1949,33 @@ func (c *Catalog) VerifyAddressToken(ctx context.Context, repository, token stri
...
+	defer it.Close()
+
+	var repos []*graveler.RepositoryRecord
+	for it.Next() {
+		record := it.Value()
+		repos = append(repos, record)
+
+	}
+	if err := it.Err(); err != nil {
+		c.log.WithError(err).Warn("failed to list repositories")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1966]: done

File: pkg/catalog/catalog.go [hunk 24]
```diff
@@ -1949,6 +1949,33 @@ func (c *Catalog) VerifyAddressToken(ctx context.Context, repository, token stri
...
+		return
+	}
+
+	for _, repo := range repos {
+		err := c.Store.DeleteExpiredAddressTokens(ctx, repo)
+		if err != nil {
+			c.log.WithError(err).WithField("repository", repo.RepositoryID).Warn("failed to delete expired address tokens")
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1973]: done

**Review by itaiad200 — APPROVED** [review]
LGTM!

**Pull request merged**